addons/survey/models/survey.py

```diff
@@ -0,0 +1,1096 @@
+        except AttributeError:
+            _logger.warning(question.question_type + ": This type of question has no validation method")
+            return {}
+        else:
+            return checker(question, post, answer_tag)
+
+    def validate_free_text(self, question, post, answer_tag):
+        errors = {}
+        answer = post[answer_tag].strip()
+        # Empty answer to mandatory question
+        if question.constr_mandatory and not answer:
+            errors.update({answer_tag: question.constr_error_msg})
+        return errors
+
+    def validate_textbox(self, question, post, answer_tag):
+        errors = {}
+        answer = post[answer_tag].strip()
+        # Empty answer to mandatory question
+        if question.constr_mandatory and not answer:
+            errors.update({answer_tag: question.constr_error_msg})
+        # Email format validation
+        # Note: this validation is very basic:
+        #     all the strings of the form
+        #     <something>@<anything>.<extension>
+        #     will be accepted
+        if answer and question.validation_email:
+            if not re.match(r"[^@]+@[^@]+\.[^@]+", answer):
+                errors.update({answer_tag: _('This answer must be an email address')})
+        # Answer validation (if properly defined)
+        # Length of the answer must be in a range
+        if answer and question.validation_required:
+            if not (question.validation_length_min <= len(answer) <= question.validation_length_max):
+                errors.update({answer_tag: question.validation_error_msg})
+        return errors
+
+    def validate_numerical_box(self, question, post, answer_tag):
+        errors = {}
+        answer = post[answer_tag].strip()
+        # Empty answer to mandatory question
+        if question.constr_mandatory and not answer:
+            errors.update({answer_tag: question.constr_error_msg})
+        # Checks if user input is a number
+        if answer:
+            try:
+                floatanswer = float(answer)
+            except ValueError:
+                errors.update({answer_tag: _('This is not a number')})
+        # Answer validation (if properly defined)
+        if answer and question.validation_required:
+            # Answer is not in the right range
+            try:
+                floatanswer = float(answer)  # check that it is a float has been done hereunder
+                if not (question.validation_min_float_value <= floatanswer <= question.validation_max_float_value):
+                    errors.update({answer_tag: question.validation_error_msg})
+            except ValueError:
+                pass
+        return errors
+
+    def validate_datetime(self, question, post, answer_tag):
+        errors = {}
+        answer = post[answer_tag].strip()
+        # Empty answer to mandatory question
+        if question.constr_mandatory and not answer:
+            errors.update({answer_tag: question.constr_error_msg})
```

Comment:
Above 3/4 lines are repetitive in validate_datetime, validate_numerical_box, validate_textbox etc methods, you define separately(But as it will be considered part of refactoring so its not an priority now)